pkg/controller/policyrecommendation/policyrecommendation_controller.go

```diff
@@ -0,0 +1,428 @@
+	rcertificatemanagement "github.com/tigera/operator/pkg/render/certificatemanagement"
+	relasticsearch "github.com/tigera/operator/pkg/render/common/elasticsearch"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const ResourceName = "policy-recommendation"
+
+var log = logf.Log.WithName("controller_policy_recommendation")
+
+// Add creates a new PolicyRecommendation Controller and adds it to the Manager. The Manager will
+// set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		// No need to start this controller
+		return nil
+	}
+	licenseAPIReady := &utils.ReadyFlag{}
+	tierWatchReady := &utils.ReadyFlag{}
+
+	reconciler := newReconciler(mgr, opts, licenseAPIReady, tierWatchReady)
+
+	controller, err := controller.New(render.PolicyRecommendationControllerName, mgr,
+		controller.Options{
+			Reconciler: reconciler,
+		})
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		log.Error(err, "Failed to establish a connection to k8s")
+		return err
+	}
+
+	go utils.WaitToAddLicenseKeyWatch(controller, k8sClient, log, licenseAPIReady)
+	go utils.WaitToAddTierWatch(networkpolicy.TigeraComponentTierName, controller, k8sClient, log, tierWatchReady)
+	go utils.WaitToAddNetworkPolicyWatches(controller, k8sClient, log, []types.NamespacedName{
+		{Name: render.PolicyRecommendationPolicyName, Namespace: render.PolicyRecommendationNamespace},
+		{Name: networkpolicy.TigeraComponentDefaultDenyPolicyName, Namespace: render.PolicyRecommendationNamespace},
+	})
+
+	return add(mgr, controller)
+}
+
+// newReconciler returns a new *reconcile.Reconciler.
+func newReconciler(
+	mgr manager.Manager,
+	opts options.AddOptions,
+	licenseAPIReady *utils.ReadyFlag,
+	tierWatchReady *utils.ReadyFlag,
+) reconcile.Reconciler {
+
+	r := &ReconcilePolicyRecommendation{
+		client:          mgr.GetClient(),
+		scheme:          mgr.GetScheme(),
+		provider:        opts.DetectedProvider,
+		status:          status.New(mgr.GetClient(), "policy-recommendation", opts.KubernetesVersion),
+		clusterDomain:   opts.ClusterDomain,
+		licenseAPIReady: licenseAPIReady,
+		tierWatchReady:  tierWatchReady,
+		usePSP:          opts.UsePSP,
+	}
+
+	r.status.Run(opts.ShutdownContext)
+
+	return r
+}
+
+// add adds watches for resources that are available at startup.
+func add(_ manager.Manager, c controller.Controller) error {
```

Comment:
this is an internal function - if we don't use the manager argument, we can just remove that argument from the signature.